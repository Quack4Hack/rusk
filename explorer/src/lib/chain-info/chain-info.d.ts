type BlockHeader = {
  date: Date;
  feespaid: number;
  hash: string;
  height: number;
  nextblockhash: string;
  prevblockhash: string;
  reward: number;
  seed: string;
  statehash: string;
};

type BlockTransactions = {
  data: Transaction[];
  stats: {
    averageGasPrice: number;
    gasLimit: number;
    gasUsed: number;
  };
};

type Block = {
  header: BlockHeader;
  transactions: BlockTransactions;
};

type ChainInfo = {
  blocks: Block[];
  transactions: Transaction[];
};

type SearchResult = {
  id: string;
  type: "block" | "transaction";
};

type Transaction = {
  blockhash: string;
  blockheight: number;
  contract: string;
  date: Date;
  feepaid: number;
  gaslimit: number;
  gasprice: number;
  gasspent: number;
  method: string;
  success: boolean;
  txerror: string;
  txid: string;
};
