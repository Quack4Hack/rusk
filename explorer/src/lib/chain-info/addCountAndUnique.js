/**
 * Returns an array of unique objects, sorted by the frequency of their occurrence,
 * with an additional count property representing the number of times
 * that city-country combination appeared in the input data.
 *
 * @param {Omit<NodeLocation, "count">[]} data
 * @returns {NodeLocation[]}
 */
function addCountAndUnique(data) {
  const countMap = data.reduce((locationMap, location) => {
    const key = `${location.city},${location.country}`;

    if (!locationMap[key]) {
      locationMap[key] = { ...location, count: 0 };
    }
    locationMap[key].count += 1;
    return locationMap;
  }, /** @type {{ [key: string]: NodeLocation }} */ ({}));

  // Return only unique city-country combinations
  return Object.values(countMap).sort((a, b) => b.count - a.count);
}

export default addCountAndUnique;
