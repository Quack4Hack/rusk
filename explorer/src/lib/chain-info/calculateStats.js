import {
  always,
  compose,
  condition,
  filterWith,
  getKey,
  partition,
} from "lamb";

import { arraySumByKey } from "$lib/dusk/array";

/** @type {(txs: Pick<GQLTransaction, "err">[]) => number} */
const getFailedTxAmount = compose(
  getKey("length"),
  filterWith((tx) => tx.err !== null)
);

/**
 * We take into account only provisioners with
 * the minimum stake amount of 1000 Dusk (1e12 Lux).
 *
 * @type {(provisioners: HostProvisioner[]) => HostProvisioner[]}
 */
const getValidProvisioners = filterWith((p) => p.amount >= 1e12);

/**
 * Sums the values of the "amount" key in the received array.
 * Returns zero if the array is empty.
 */
const sumByAmount = condition(
  (provisioners) => provisioners.length > 0,
  arraySumByKey("amount"),
  always(0)
);

/**
 * @param {HostProvisioner[]} provisioners
 * @param {number} lastHeight
 * @param {Pick<GQLTransaction, "err">[]} last100BlocksTxs
 * @returns {Stats}
 */
function calculateStats(provisioners, lastHeight, last100BlocksTxs) {
  const [activeProvisioners, waitingProvisioners] = partition(
    getValidProvisioners(provisioners),
    (p) => p.eligibility <= lastHeight
  );

  return {
    activeProvisioners: activeProvisioners.length,
    activeStake: sumByAmount(activeProvisioners),
    lastBlock: lastHeight,
    txs100blocks: {
      failed: getFailedTxAmount(last100BlocksTxs),
      transfers: last100BlocksTxs.length,
    },
    waitingProvisioners: waitingProvisioners.length,
    waitingStake: sumByAmount(waitingProvisioners),
  };
}

export default calculateStats;
