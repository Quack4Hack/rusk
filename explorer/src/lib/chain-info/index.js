export { default as transformBlock } from "./transformBlock";
export { default as transformTransaction } from "./transformTransaction";
