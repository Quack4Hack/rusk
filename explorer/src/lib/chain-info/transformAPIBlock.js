import { mapWith, pipe, skip, updateIn } from "lamb";

import { unixTsToDate } from "$lib/dusk/date";

import { transformAPITransaction } from ".";

/** @type {(v: APIBlockHeader) => Required<APIBlockHeader>} */
const mergeWithDefaults = (v) => ({
  nextblockhash: "",
  prevblockhash: "",
  statehash: "",
  ...v,
});

/** @type {(header: Required<APIBlockHeader>) => BlockHeader} */
const addHeaderDate = (header) => ({
  ...header,
  date: unixTsToDate(header.ts),
});

/** @type {(v: APIBlockHeader) => BlockHeader} */
const transformAPIBlockHeader = pipe([
  mergeWithDefaults,
  addHeaderDate,
  skip(["__typename", "timestamp", "ts", "version"]),
]);

/** @type {(v: APIBlock) => Block} */
const transformAPIBlock = ({ header, transactions }) => ({
  header: transformAPIBlockHeader(header),
  transactions: updateIn(
    transactions,
    "data",
    mapWith(transformAPITransaction)
  ),
});

export default transformAPIBlock;
