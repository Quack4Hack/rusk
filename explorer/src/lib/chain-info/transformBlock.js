import { mapWith, pipe, skip, updateIn } from "lamb";

import { unixTsToDate } from "$lib/dusk/date";

import { transformTransaction } from ".";

/** @type {(v: APIBlockHeader) => Required<APIBlockHeader>} */
const mergeWithDefaults = (v) => ({
  nextblockhash: "",
  prevblockhash: "",
  statehash: "",
  ...v,
});

/** @type {(header: Required<APIBlockHeader>) => BlockHeader} */
const addHeaderDate = (header) => ({
  ...header,
  date: unixTsToDate(header.ts),
});

/** @type {(v: APIBlockHeader) => BlockHeader} */
const transformBlockHeader = pipe([
  mergeWithDefaults,
  addHeaderDate,
  skip(["__typename", "timestamp", "ts"]),
]);

/** @type {(v: APIBlock) => Block} */
const transformBlock = ({ header, transactions }) => ({
  header: transformBlockHeader(header),
  transactions: updateIn(transactions, "data", mapWith(transformTransaction)),
});

export default transformBlock;
