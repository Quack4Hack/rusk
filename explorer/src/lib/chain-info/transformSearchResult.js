import { compose, getPathIn, mapWith } from "lamb";

/**
 * @template {Record<PropertyKey, any>} T
 * @template {string} U
 * @param {T} source
 * @param {U} path
 * @returns {Exclude<import("lamb").GetPath<T, U>, undefined> | []}
 */
const getPathOrEmptyArray = (source, path) => getPathIn(source, path) ?? [];

/** @type {(v: APISearchResult) => SearchResult[]} */
const transformSearchResult = compose(
  mapWith((entry) =>
    entry?.header?.hash
      ? { id: entry.header.hash, type: "block" }
      : { id: entry.txid, type: "transaction" }
  ),
  (v) => [
    ...getPathOrEmptyArray(v, "data.data.blocks"),
    ...getPathOrEmptyArray(v, "data.data.transactions"),
  ]
);

export default transformSearchResult;
