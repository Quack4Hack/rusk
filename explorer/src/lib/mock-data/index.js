export { default as apiMarketData } from "./api-market-data.json";
export { default as apiNodeLocations } from "./api-node-locations.json";
export { default as apiStats } from "./api-stats.json";
export { default as gqlBlock } from "./gql-block.json";
export { default as gqlBlocks } from "./gql-blocks.json";
export { default as gqlLatestChainInfo } from "./gql-latest-chain-info.json";
export { default as gqlSearchPossibleResults } from "./gql-search-possible-results.json";
export { default as gqlTransaction } from "./gql-transaction.json";
export { default as gqlTransactionDetails } from "./gql-transaction-details.json";
export { default as gqlTransactions } from "./gql-transactions.json";
