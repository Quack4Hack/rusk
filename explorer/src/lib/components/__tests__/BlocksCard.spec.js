import { afterAll, afterEach, describe, expect, it, vi } from "vitest";
import { cleanup, fireEvent, render } from "@testing-library/svelte";
import { apiBlocks } from "$lib/mock-data";
import { transformBlock } from "$lib/chain-info";
import { BlocksCard } from "..";
import { mapWith, slice } from "lamb";

const transformBlocks = mapWith(transformBlock);
const data = slice(transformBlocks(apiBlocks.data.blocks), 0, 10);

describe("Blocks Card", () => {
  vi.useFakeTimers();
  vi.setSystemTime(new Date(2024, 4, 20));
  const baseProps = {
    blocks: data,
    error: null,
    loading: false,
  };
  const baseOptions = {
    props: baseProps,
    target: document.body,
  };

  afterEach(cleanup);

  afterAll(() => {
    vi.useRealTimers();
  });

  it("should render the `BlocksCard` component", () => {
    const { container } = render(BlocksCard, baseOptions);

    expect(container.firstChild).toMatchSnapshot();
  });

  it("should disable the `Show More` button is the card is in the loading state", () => {
    const loading = true;

    const { container, getByRole } = render(BlocksCard, {
      ...baseOptions,
      props: { ...baseProps, loading },
    });

    expect(getByRole("button")).toBeDisabled();

    expect(container.firstChild).toMatchSnapshot();
  });

  it("should disable the `Show More` button if there is no more data to display", async () => {
    const loading = true;

    const { container, getByRole } = render(BlocksCard, {
      ...baseOptions,
      props: { ...baseProps, loading },
    });

    const button = getByRole("button");

    const showMoreIncrement = 15;

    const clicks = Math.ceil(data.length / showMoreIncrement) - 1;

    Array.from({ length: clicks }).forEach(async () => {
      await fireEvent.click(button);
    });

    expect(button).toBeDisabled();

    expect(container.firstChild).toMatchSnapshot();
  });
});
