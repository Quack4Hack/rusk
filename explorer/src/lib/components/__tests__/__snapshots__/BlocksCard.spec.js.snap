// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Blocks Card > should disable the \`Show More\` button if there is no more data to display 1`] = `
<div
  class="dusk-card data-card"
>
  <header
    class="data-card__header"
    slot="header"
  >
    <h1
      class="data-card__header-title"
    >
      Blocks
    </h1>
     
    <button
      class="dusk-button dusk-button--type--button dusk-button--variant--secondary dusk-button--size--normal"
      disabled=""
      type="button"
    >
      <span
        class="dusk-button__text"
      >
        Show More
      </span>
    </button>
    <!--&lt;Button&gt;-->
  </header>
   
  <div
    class="table-container blocks-table blocks-card__table"
  >
    <table
      class="table"
    >
      <thead
        class="table__head"
      >
        <tr
          class="table__row"
        >
          <th
            class="table__header-cell"
          >
            # Block
          </th>
          
          <!--&lt;TableCell&gt;-->
           
          <th
            class="table__header-cell"
          >
            Fee
          </th>
          
          <!--&lt;TableCell&gt;-->
           
          <th
            class="table__header-cell"
          >
            Txn(s)
          </th>
          
          <!--&lt;TableCell&gt;-->
           
          <th
            class="table__header-cell"
          >
            Rewards
          </th>
          
          <!--&lt;TableCell&gt;-->
        </tr>
        <!--&lt;TableRow&gt;-->
      </thead>
      <!--&lt;TableHead&gt;-->
       
      <tbody
        class="table__body"
      >
        
        <tr
          class="table__row"
        >
          <td
            class="table__data-cell"
          >
            <a
              class="dusk-anchor block__link"
              href="/some-base-path/blocks/block?id=9ebf7e785d75bc7c8fdc386f6806ffa0964b19c6f18f56d869274ea046ed0148"
            >
              487,491
            </a>
            <!--&lt;Anchor&gt;-->
            <!--&lt;AppAnchor&gt;-->
             
            <small
              class="block__time"
            >
              last month
            </small>
          </td>
          
          <!--&lt;TableCell&gt;-->
           
          <td
            class="table__data-cell"
          >
            <b
              class="block__fee-avg-label"
            >
              AVG:
            </b>
             
            0
            <br />
             
            <b
              class="block__fee-total-label"
            >
              TOTAL:
            </b>
             
            0
          </td>
          
          <!--&lt;TableCell&gt;-->
           
          <td
            class="table__data-cell"
          >
            0
          </td>
          
          <!--&lt;TableCell&gt;-->
           
          <td
            class="table__data-cell"
          >
            <span
              class="dusk-badge dusk-badge--variant-alt"
            >
              16 Dusk
            </span>
            <!--&lt;Badge&gt;-->
          </td>
          
          <!--&lt;TableCell&gt;-->
           
        </tr>
        <!--&lt;TableRow&gt;-->
        
        <tr
          class="table__row"
        >
          <td
            class="table__data-cell"
          >
            <a
              class="dusk-anchor block__link"
              href="/some-base-path/blocks/block?id=d5d5d211791468e09d3224eb194b9ae89f0ba1ad5dea1e8e458e7a3cb9b01bfa"
            >
              487,490
            </a>
            <!--&lt;Anchor&gt;-->
            <!--&lt;AppAnchor&gt;-->
             
            <small
              class="block__time"
            >
              last month
            </small>
          </td>
          
          <!--&lt;TableCell&gt;-->
           
          <td
            class="table__data-cell"
          >
            <b
              class="block__fee-avg-label"
            >
              AVG:
            </b>
             
            0
            <br />
             
            <b
              class="block__fee-total-label"
            >
              TOTAL:
            </b>
             
            0
          </td>
          
          <!--&lt;TableCell&gt;-->
           
          <td
            class="table__data-cell"
          >
            0
          </td>
          
          <!--&lt;TableCell&gt;-->
           
          <td
            class="table__data-cell"
          >
            <span
              class="dusk-badge dusk-badge--variant-alt"
            >
              16 Dusk
            </span>
            <!--&lt;Badge&gt;-->
          </td>
          
          <!--&lt;TableCell&gt;-->
           
        </tr>
        <!--&lt;TableRow&gt;-->
        
        <tr
          class="table__row"
        >
          <td
            class="table__data-cell"
          >
            <a
              class="dusk-anchor block__link"
              href="/some-base-path/blocks/block?id=07d8c3fedd76afcf108dacc311867062393a39cf34a76d5226c7ec8360017117"
            >
              487,489
            </a>
            <!--&lt;Anchor&gt;-->
            <!--&lt;AppAnchor&gt;-->
             
            <small
              class="block__time"
            >
              last month
            </small>
          </td>
          
          <!--&lt;TableCell&gt;-->
           
          <td
            class="table__data-cell"
          >
            <b
              class="block__fee-avg-label"
            >
              AVG:
            </b>
             
            0
            <br />
             
            <b
              class="block__fee-total-label"
            >
              TOTAL:
            </b>
             
            0
          </td>
          
          <!--&lt;TableCell&gt;-->
           
          <td
            class="table__data-cell"
          >
            0
          </td>
          
          <!--&lt;TableCell&gt;-->
           
          <td
            class="table__data-cell"
          >
            <span
              class="dusk-badge dusk-badge--variant-alt"
            >
              16 Dusk
            </span>
            <!--&lt;Badge&gt;-->
          </td>
          
          <!--&lt;TableCell&gt;-->
           
        </tr>
        <!--&lt;TableRow&gt;-->
        
        <tr
          class="table__row"
        >
          <td
            class="table__data-cell"
          >
            <a
              class="dusk-anchor block__link"
              href="/some-base-path/blocks/block?id=6ffed76cc339260d215219c3e6a710ae8fc0d81f68cfb201b1db1c9c57d9ca4c"
            >
              487,488
            </a>
            <!--&lt;Anchor&gt;-->
            <!--&lt;AppAnchor&gt;-->
             
            <small
              class="block__time"
            >
              last month
            </small>
          </td>
          
          <!--&lt;TableCell&gt;-->
           
          <td
            class="table__data-cell"
          >
            <b
              class="block__fee-avg-label"
            >
              AVG:
            </b>
             
            0
            <br />
             
            <b
              class="block__fee-total-label"
            >
              TOTAL:
            </b>
             
            0
          </td>
          
          <!--&lt;TableCell&gt;-->
           
          <td
            class="table__data-cell"
          >
            0
          </td>
          
          <!--&lt;TableCell&gt;-->
           
          <td
            class="table__data-cell"
          >
            <span
              class="dusk-badge dusk-badge--variant-alt"
            >
              16 Dusk
            </span>
            <!--&lt;Badge&gt;-->
          </td>
          
          <!--&lt;TableCell&gt;-->
           
        </tr>
        <!--&lt;TableRow&gt;-->
        
        <tr
          class="table__row"
        >
          <td
            class="table__data-cell"
          >
            <a
              class="dusk-anchor block__link"
              href="/some-base-path/blocks/block?id=284f14e50099db32bc34fc0cbecc016224100dd461fa4e2e02a60ffc77f7b391"
            >
              487,487
            </a>
            <!--&lt;Anchor&gt;-->
            <!--&lt;AppAnchor&gt;-->
             
            <small
              class="block__time"
            >
              last month
            </small>
          </td>
          
          <!--&lt;TableCell&gt;-->
           
          <td
            class="table__data-cell"
          >
            <b
              class="block__fee-avg-label"
            >
              AVG:
            </b>
             
            0
            <br />
             
            <b
              class="block__fee-total-label"
            >
              TOTAL:
            </b>
             
            0
          </td>
          
          <!--&lt;TableCell&gt;-->
           
          <td
            class="table__data-cell"
          >
            0
          </td>
          
          <!--&lt;TableCell&gt;-->
           
          <td
            class="table__data-cell"
          >
            <span
              class="dusk-badge dusk-badge--variant-alt"
            >
              16 Dusk
            </span>
            <!--&lt;Badge&gt;-->
          </td>
          
          <!--&lt;TableCell&gt;-->
           
        </tr>
        <!--&lt;TableRow&gt;-->
        
        <tr
          class="table__row"
        >
          <td
            class="table__data-cell"
          >
            <a
              class="dusk-anchor block__link"
              href="/some-base-path/blocks/block?id=adb7e41bd1509ee542b782a9ae59d82aec00d7ac0810b0bf6d6a6ea6fb54bced"
            >
              487,486
            </a>
            <!--&lt;Anchor&gt;-->
            <!--&lt;AppAnchor&gt;-->
             
            <small
              class="block__time"
            >
              last month
            </small>
          </td>
          
          <!--&lt;TableCell&gt;-->
           
          <td
            class="table__data-cell"
          >
            <b
              class="block__fee-avg-label"
            >
              AVG:
            </b>
             
            0
            <br />
             
            <b
              class="block__fee-total-label"
            >
              TOTAL:
            </b>
             
            0
          </td>
          
          <!--&lt;TableCell&gt;-->
           
          <td
            class="table__data-cell"
          >
            0
          </td>
          
          <!--&lt;TableCell&gt;-->
           
          <td
            class="table__data-cell"
          >
            <span
              class="dusk-badge dusk-badge--variant-alt"
            >
              16 Dusk
            </span>
            <!--&lt;Badge&gt;-->
          </td>
          
          <!--&lt;TableCell&gt;-->
           
        </tr>
        <!--&lt;TableRow&gt;-->
        
        <tr
          class="table__row"
        >
          <td
            class="table__data-cell"
          >
            <a
              class="dusk-anchor block__link"
              href="/some-base-path/blocks/block?id=b16fa4cac5f70b0b021fb7fc0f5388c6a4ad56d21f5170dea8e2abdfc836462c"
            >
              487,485
            </a>
            <!--&lt;Anchor&gt;-->
            <!--&lt;AppAnchor&gt;-->
             
            <small
              class="block__time"
            >
              last month
            </small>
          </td>
          
          <!--&lt;TableCell&gt;-->
           
          <td
            class="table__data-cell"
          >
            <b
              class="block__fee-avg-label"
            >
              AVG:
            </b>
             
            0
            <br />
             
            <b
              class="block__fee-total-label"
            >
              TOTAL:
            </b>
             
            0
          </td>
          
          <!--&lt;TableCell&gt;-->
           
          <td
            class="table__data-cell"
          >
            0
          </td>
          
          <!--&lt;TableCell&gt;-->
           
          <td
            class="table__data-cell"
          >
            <span
              class="dusk-badge dusk-badge--variant-alt"
            >
              16 Dusk
            </span>
            <!--&lt;Badge&gt;-->
          </td>
          
          <!--&lt;TableCell&gt;-->
           
        </tr>
        <!--&lt;TableRow&gt;-->
        
        <tr
          class="table__row"
        >
          <td
            class="table__data-cell"
          >
            <a
              class="dusk-anchor block__link"
              href="/some-base-path/blocks/block?id=0848155ae84398b4a2424a526d9a9e28befa2fff5e54eb253592df9e62cda572"
            >
              487,484
            </a>
            <!--&lt;Anchor&gt;-->
            <!--&lt;AppAnchor&gt;-->
             
            <small
              class="block__time"
            >
              last month
            </small>
          </td>
          
          <!--&lt;TableCell&gt;-->
           
          <td
            class="table__data-cell"
          >
            <b
              class="block__fee-avg-label"
            >
              AVG:
            </b>
             
            0
            <br />
             
            <b
              class="block__fee-total-label"
            >
              TOTAL:
            </b>
             
            0
          </td>
          
          <!--&lt;TableCell&gt;-->
           
          <td
            class="table__data-cell"
          >
            0
          </td>
          
          <!--&lt;TableCell&gt;-->
           
          <td
            class="table__data-cell"
          >
            <span
              class="dusk-badge dusk-badge--variant-alt"
            >
              16 Dusk
            </span>
            <!--&lt;Badge&gt;-->
          </td>
          
          <!--&lt;TableCell&gt;-->
           
        </tr>
        <!--&lt;TableRow&gt;-->
        
        <tr
          class="table__row"
        >
          <td
            class="table__data-cell"
          >
            <a
              class="dusk-anchor block__link"
              href="/some-base-path/blocks/block?id=e583705447034015672243dc5ec694e673344d3969846d5606cca88e2a6fef9f"
            >
              487,483
            </a>
            <!--&lt;Anchor&gt;-->
            <!--&lt;AppAnchor&gt;-->
             
            <small
              class="block__time"
            >
              last month
            </small>
          </td>
          
          <!--&lt;TableCell&gt;-->
           
          <td
            class="table__data-cell"
          >
            <b
              class="block__fee-avg-label"
            >
              AVG:
            </b>
             
            0
            <br />
             
            <b
              class="block__fee-total-label"
            >
              TOTAL:
            </b>
             
            0
          </td>
          
          <!--&lt;TableCell&gt;-->
           
          <td
            class="table__data-cell"
          >
            0
          </td>
          
          <!--&lt;TableCell&gt;-->
           
          <td
            class="table__data-cell"
          >
            <span
              class="dusk-badge dusk-badge--variant-alt"
            >
              16 Dusk
            </span>
            <!--&lt;Badge&gt;-->
          </td>
          
          <!--&lt;TableCell&gt;-->
           
        </tr>
        <!--&lt;TableRow&gt;-->
        
        <tr
          class="table__row"
        >
          <td
            class="table__data-cell"
          >
            <a
              class="dusk-anchor block__link"
              href="/some-base-path/blocks/block?id=8bc650bb577f6d4a93b41523712f2f2c82911395e871c022dfa56d54c12061bc"
            >
              487,482
            </a>
            <!--&lt;Anchor&gt;-->
            <!--&lt;AppAnchor&gt;-->
             
            <small
              class="block__time"
            >
              last month
            </small>
          </td>
          
          <!--&lt;TableCell&gt;-->
           
          <td
            class="table__data-cell"
          >
            <b
              class="block__fee-avg-label"
            >
              AVG:
            </b>
             
            0
            <br />
             
            <b
              class="block__fee-total-label"
            >
              TOTAL:
            </b>
             
            0
          </td>
          
          <!--&lt;TableCell&gt;-->
           
          <td
            class="table__data-cell"
          >
            0
          </td>
          
          <!--&lt;TableCell&gt;-->
           
          <td
            class="table__data-cell"
          >
            <span
              class="dusk-badge dusk-badge--variant-alt"
            >
              16 Dusk
            </span>
            <!--&lt;Badge&gt;-->
          </td>
          
          <!--&lt;TableCell&gt;-->
           
        </tr>
        <!--&lt;TableRow&gt;-->
        
      </tbody>
      <!--&lt;TableBody&gt;-->
    </table>
  </div>
  <!--&lt;Table&gt;-->
  <!--&lt;BlocksTable&gt;-->
  
  
   
</div>
`;

exports[`Blocks Card > should disable the \`Show More\` button is the card is in the loading state 1`] = `
<div
  class="dusk-card data-card"
>
  <header
    class="data-card__header"
    slot="header"
  >
    <h1
      class="data-card__header-title"
    >
      Blocks
    </h1>
     
    <button
      class="dusk-button dusk-button--type--button dusk-button--variant--secondary dusk-button--size--normal"
      disabled=""
      type="button"
    >
      <span
        class="dusk-button__text"
      >
        Show More
      </span>
    </button>
    <!--&lt;Button&gt;-->
  </header>
   
  <p
    class="data-card__loading-notice"
  >
    Loading...
  </p>
  
   
</div>
`;

exports[`Blocks Card > should render the \`BlocksCard\` component 1`] = `
<div
  class="dusk-card data-card"
>
  <header
    class="data-card__header"
    slot="header"
  >
    <h1
      class="data-card__header-title"
    >
      Blocks
    </h1>
     
    <button
      class="dusk-button dusk-button--type--button dusk-button--variant--secondary dusk-button--size--normal"
      type="button"
    >
      <span
        class="dusk-button__text"
      >
        Show More
      </span>
    </button>
    <!--&lt;Button&gt;-->
  </header>
   
  
   
</div>
`;
